rustls/src/msgs/deframer/outer.rs

```diff
@@ -0,0 +1,236 @@
+use core::mem;
+
+use crate::error::{Error, InvalidMessage};
+use crate::msgs::codec::Reader;
+use crate::msgs::message::{
+    read_opaque_message_header, InboundOpaqueMessage, MessageError, HEADER_SIZE,
+};
+
+/// A deframer of outer TLS messages.
+#[derive(Default)]
+pub(crate) struct OuterDeframer {
+    /// Set if the peer is not talking TLS, but some other
+    /// protocol.  The caller should abort the connection, because
+    /// the deframer cannot recover.
+    last_error: Option<Error>,
+}
+
+impl<'a> OuterDeframer {
+    pub(crate) fn deframe<'b>(&'a mut self, buf: &'b mut [u8]) -> DeframerIter<'a, 'b> {
+        DeframerIter {
+            deframer: self,
+            buf,
+            consumed: 0,
+        }
+    }
+
+    fn fuse_error(&mut self, err: impl Into<Error>) -> Error {
+        let err = err.into();
+        self.last_error = Some(err.clone());
+        err
+    }
+}
+
+pub(crate) struct DeframerIter<'a, 'b> {
+    deframer: &'a mut OuterDeframer,
+    buf: &'b mut [u8],
+    consumed: usize,
+}
+
+impl DeframerIter<'_, '_> {
+    pub(crate) fn bytes_consumed(&self) -> usize {
+        self.consumed
+    }
+}
+
+impl<'a, 'b> Iterator for DeframerIter<'a, 'b> {
```

Comment:
I guess this is similar as the point that @cpu made before, but (perhaps more concretely) this would be easier to review if it was introduced in the same commit that removed the original version of this.

Reply:
^ Have done this.